README.md
# select_form_field

[![pub package](https://img.shields.io/pub/v/select_form_field.svg)](https://pub.dartlang.org/packages/select_form_field)

<a href="https://www.buymeacoffee.com/hslbetto" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-blue.png" alt="Buy Me A Beer" style="width: 150px !important;"></a>

A Flutter select field widget. It shows a list of options in a dropdown menu.\
This widget extend TextField and has a similar behavior as TextFormField

## Usage

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  ...
  select_form_field: "^2.2.0"
```

In your library add the following import:

```dart
import 'package:select_form_field/select_form_field.dart';
```

For help getting started with Flutter, view the online [documentation](https://flutter.io/).

## Example

Set items using a List map passing:
* `value`: [String], 
* `textStyle`: [TextStyle | null],
* `label`: [String | null], 
* `icon`: [Widget | null],
* `enable`: [bool | null],

``` dart
final List<Map<String, dynamic>> _items = [
  {
    'value': 'boxValue',
    'label': 'Box Label',
    'icon': Icon(Icons.stop),
  },
  {
    'value': 'circleValue',
    'label': 'Circle Label',
    'icon': Icon(Icons.fiber_manual_record),
    'textStyle': TextStyle(color: Colors.red),
  },
  {
    'value': 'starValue',
    'label': 'Star Label',
    'enable': false,
    'icon': Icon(Icons.grade),
  },
];
```

``` dart
SelectFormField(
  type: SelectFormFieldType.dropdown, // or can be dialog
  initialValue: 'circle',
  icon: Icon(Icons.format_shapes),
  labelText: 'Shape',
  items: _items,
  onChanged: (val) => print(val),
  onSaved: (val) => print(val),
);
```

The result of val in `onChanged`, `validator` and `onSaved` will be a String.\
So, if you tap on Box Label item on select menu the result will be `boxValue`.

## Preview
![Overview](https://raw.githubusercontent.com/m3uzz/select_form_field/master/doc/images/select_form_field.gif)


CHANGELOG.md
## 2.2.0 - 2021-07-27

* Change how SelectFormField identify and set the initial value from item list (issue #8).
* Fixed field cleaning (issue #11).

## 2.1.0 - 2021-07-07

* Fixed bug of return when click outside the select items.
* Improved validation of initialValue and items parameters.

## 2.0.0 - 2021-03-09

* Migrate to null safety.
* Update SDK constraints to >=2.12.0 <3.0.0 based on beta release guidelines.
* Remove deprecated FlatButton and maxLengthEnforced.

## 1.1.0 - 2020-12-07

* Increment an option to show itens as dropdown or dialog with search.
* Exemple update.

## 1.0.1 - 2020-08-05

* Fixed issue #1 - text render overflowed.

## 1.0.0 - 2020-07-17

* Fixed use of widget with empty initial value.
* Improved README.md.
* Improved testing.
* Improved example app.
* Final validation.

## 0.1.1 - 2020-07-16

* Example page published.
  
## 0.1.0 - 2020-07-16

* Initial Open Source release.

test/select_form_field_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:select_form_field/select_form_field.dart';

void main() {
  testWidgets('Testing instantiate SelectFormField',
      (WidgetTester tester) async {
    var myWidget = MyWidget();
    await tester.pumpWidget(myWidget);

    expect(
      find.text('test'),
      findsOneWidget,
      reason: 'SelectFormField value not found!',
    );
    var selectField = find.byType(SelectFormField);
    expect(
      selectField,
      findsOneWidget,
      reason: 'SelectFormField not found!',
    );
    expect(
      find.text('Circle Label'),
      findsOneWidget,
      reason: 'SelectFormField initial label value not found!',
    );

    await tester.tap(selectField);
    await tester.pumpAndSettle();

    var boxLabel = find.text('Box Label');
    expect(
      boxLabel,
      findsOneWidget,
      reason: 'SelectFormField Box option not displayed on dialog!',
    );
    expect(
      find.text('Star Label'),
      findsOneWidget,
      reason: 'SelectFormField Star option not displayed on dialog!',
    );

    var item = find.ancestor(
      of: boxLabel,
      matching: find.byType(InkWell),
    );
    expect(
      item,
      findsOneWidget,
      reason: 'SelectFormField InkWell not found!',
    );

    await tester.tap(item);
    await tester.pumpAndSettle();

    expect(
      find.text('Box Label'),
      findsOneWidget,
      reason: 'SelectFormField Box not selected!',
    );
    expect(
      find.text('boxValue'),
      findsOneWidget,
      reason: 'SelectFormField value result not changed!',
    );
  });
}

class MyWidget extends StatefulWidget {
  MyWidget({Key? key}) : super(key: key);

  @override
  _MyWidgetState createState() => _MyWidgetState();
}

class _MyWidgetState extends State<MyWidget> {
  String _value = 'test';

  final List<Map<String, dynamic>> _items = [
    {
      'value': 'boxValue',
      'label': 'Box Label',
      'icon': Icon(Icons.stop),
    },
    {
      'value': 'circleValue',
      'label': 'Circle Label',
      'icon': Icon(Icons.fiber_manual_record),
      'textStyle': TextStyle(color: Colors.red),
    },
    {
      'value': 'starValue',
      'label': 'Star Label',
      'enable': false,
      'icon': Icon(Icons.grade),
    },
  ];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Column(
          children: [
            SelectFormField(
              initialValue: 'circleValue',
              icon: Icon(Icons.format_shapes),
              labelText: 'Shape',
              items: _items,
              onChanged: (val) => setState(() => _value = val),
            ),
            Text(_value),
          ],
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:select_form_field/select_form_field.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter SelectFormField Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  GlobalKey<FormState> _oFormKey = GlobalKey<FormState>();
  TextEditingController? _controller;
  //String _initialValue;
  String _valueChanged = '';
  String _valueToValidate = '';
  String _valueSaved = '';

  final List<Map<String, dynamic>> _items = [
    {
      'value': 'boxValue',
      'label': 'Box Label',
      'icon': Icon(Icons.stop),
    },
    {
      'value': 'circleValue',
      'label': 'Circle Label Loooooooooooooooooooong text',
      'icon': Icon(Icons.fiber_manual_record),
      'textStyle': TextStyle(color: Colors.red),
    },
    {
      'value': 'starValue',
      'label': 'Star Label',
      'enable': false,
      'icon': Icon(Icons.grade),
    },
  ];

  /*final List<Map<String, dynamic>> _items = [
    {'value': 1},
    {'value': 2},
    {'value': 3},
    {'value': 4},
    {'value': 5},
    {'value': 6},
    {'value': 7},
    {'value': 8},
    {'value': 9},
    {'value': 10},
    {'value': 11},
    {'value': 12},
    {'value': 13},
    {'value': 14},
    {'value': 15},
    {'value': 16},
    {'value': 17},
    {'value': 18},
    {'value': 19},
    {'value': 20},
    {'value': 21},
    {'value': 22},
    {'value': 23},
    {'value': 24},
    {'value': 25},
    {'value': 26},
    {'value': 27},
    {'value': 28},
    {'value': 29},
    {'value': 30},
    {'value': 31},
  ];*/

  @override
  void initState() {
    super.initState();

    //_initialValue = 'starValue';
    _controller = TextEditingController(text: '2');

    _getValue();
  }

  /// This implementation is just to simulate a load data behavior
  /// from a data base sqlite or from a API
  Future<void> _getValue() async {
    await Future.delayed(const Duration(seconds: 3), () {
      setState(() {
        //_initialValue = 'circleValue';
        _controller?.text = 'circleValue';
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Flutter SelectFormField Demo'),
      ),
      body: SingleChildScrollView(
        padding: EdgeInsets.only(left: 20, right: 20, top: 10),
        child: Form(
          key: _oFormKey,
          child: Column(
            children: <Widget>[
              SelectFormField(
                type: SelectFormFieldType.dialog,
                controller: _controller,
                //initialValue: _initialValue,
                icon: Icon(Icons.format_shapes),
                labelText: 'Shape',
                changeIcon: true,
                dialogTitle: 'Pick a item',
                dialogCancelBtn: 'CANCEL',
                enableSearch: true,
                dialogSearchHint: 'Search item',
                items: _items,
                onChanged: (val) => setState(() => _valueChanged = val),
                validator: (val) {
                  setState(() => _valueToValidate = val ?? '');
                  return null;
                },
                onSaved: (val) => setState(() => _valueSaved = val ?? ''),
              ),
              SizedBox(height: 30),
              Text(
                'SelectFormField data value onChanged:',
                style: TextStyle(fontWeight: FontWeight.bold),
              ),
              SizedBox(height: 10),
              SelectableText(_valueChanged),
              SizedBox(height: 30),
              ElevatedButton(
                onPressed: () {
                  final loForm = _oFormKey.currentState;

                  if (loForm?.validate() == true) {
                    loForm?.save();
                  }
                },
                child: Text('Submit'),
              ),
              SizedBox(height: 30),
              Text(
                'SelectFormField data value validator:',
                style: TextStyle(fontWeight: FontWeight.bold),
              ),
              SizedBox(height: 10),
              SelectableText(_valueToValidate),
              SizedBox(height: 30),
              Text(
                'SelectFormField data value onSaved:',
                style: TextStyle(fontWeight: FontWeight.bold),
              ),
              SizedBox(height: 10),
              SelectableText(_valueSaved),
              SizedBox(height: 30),
              ElevatedButton(
                onPressed: () {
                  final loForm = _oFormKey.currentState;
                  loForm?.reset();

                  setState(() {
                    _valueChanged = '';
                    _valueToValidate = '';
                    _valueSaved = '';
                    _controller?.clear();
                  });
                },
                child: Text('Reset'),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
